speech_to_text/speech_to_text.livemd

```diff
@@ -0,0 +1,217 @@
+
+You will see how to fetch the audio from your microphone, perform preprocessing and create your own Membrane element that runs an AI speech-to-text conversion model.
+
+## Element performing a speech to text
+
+We need to write a custom sink that will perform speech-to-text and display the resulting transcription. Let's call this element `SpeechToText`.
+
+To perform the transcription the element will use the Open AI's [Whisper](https://openai.com/research/whisper) model. 
+It can be easily loaded and used with a "little" help from [Bumblebee](https://github.com/elixir-nx/bumblebee).
+
+The Whipser model requires the input audio samples to be in `f32le` format, which means, that they are represented as floating numbers written on 32 bits, with little endian bytes order. The required input sample rate is 16 kHz, as well as only single-channel samples are allowed.
+
+### Initialization
+
+In the initialization process, we load the `Whisper` model, along with the featurizer, tokenizer, and generation configuration. Then we create a speech-to-text serving and indicate that we want to use `EXLA` backend for the tensor operations.
+
+### Buffers handling
+
+Once a buffer arrives, we check if it contains enough amount of audio and accumulate it in the state otherwise. Then we perform basic VAD (Voice Activity Detection) to remove silent chunks, what improves the model behaviour. Then we convert the audio to an Nx tensor, feed the model and parse output, which is then sent to via the output pad.
+
+### Problems and potential improvements
+
+The main problem with our approach is that we feed the model with small chunks, so that it lacks the context of what is before and after. If a chunk starts or ends in a middle of a word, it's very hard for the model to recognize it properly. The solution for that is to make the chunks overlap and then apply a clever merging algorithm. This is almost implemented in the Bumblebee - see https://github.com/elixir-nx/bumblebee/issues/261. Merging could be also done outside of Bumblebee, but it's hard to say how it would perform, since it would have to rely on the output text or timestamps instead of tokens. Alternatively, we could apply some clever algorithm that would tell us where to split the chunks basing on silent breaks.
+
+Another issue is that Whisper seems to hallucinate when given a noise or silence. We're partially solving that by removing the silence whenever possible. The silence detection and removal is not a trivial task and can possibly be improved. Also, applying noise reduction could be helpful.
+
+We welcome ideas and contributions addressing these problems or bringing other improvements.
+
+```elixir
+defmodule SpeechToText do
+  use Membrane.Filter
+
+  alias Membrane.RawAudio
+  require Membrane.Logger
+
+  # 8000 4-byte samples, at 16kHz it's half a second
```

Comment:
How about setting the time (in this case 500ms) as an attribute and calculating the `vad_chunk_size` based on that? (I believe we can use (that)[https://github.com/membraneframework/membrane_raw_audio_format/blob/49d5dd11717d872edb377aa5afa26f2392225b7c/lib/membrane_raw_audio.ex#L353) function for this purpose)